pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultiStageQuerySemaphore.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.requesthandler;
+
+import com.google.common.base.Preconditions;
+import java.util.Collections;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.HelixConfigScope;
+import org.apache.helix.model.builder.HelixConfigScopeBuilder;
+import org.apache.pinot.broker.broker.helix.ClusterChangeHandler;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class helps limit the number of multi-stage queries being executed concurrently. Currently, this limit is
+ * applied at the broker level, but could be moved to the server level in the future.
+ */
+public class MultiStageQuerySemaphore extends Semaphore implements ClusterChangeHandler {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MultiStageQuerySemaphore.class);
+
+  private HelixManager _helixManager;
+  private HelixAdmin _helixAdmin;
+  private HelixConfigScope _helixConfigScope;
+  private int _numBrokers;
+  private int _maxConcurrentQueries;
+  private int _totalPermits;
+
+  public MultiStageQuerySemaphore(int numBrokers, int maxConcurrentQueries) {
+    super(Math.max(1, maxConcurrentQueries / Math.max(numBrokers, 1)), true);
+    _maxConcurrentQueries = maxConcurrentQueries;
+    _numBrokers = Math.max(1, numBrokers);
+    _totalPermits =
+        maxConcurrentQueries > 0 ? Math.max(1, maxConcurrentQueries / Math.max(numBrokers, 1)) : maxConcurrentQueries;
+  }
```

Comment:
I think we are using too many _max_es here. IMO it would be better to create a static factory method that does all the max logic and then call a simple constructor.